File: src/tui/theme.zig
const std = @import("std");

pub const Rgb = [3]u8;

pub const Theme = struct {
    name: []const u8,
    background: Rgb,
    foreground: Rgb,
    primary: Rgb,
    accent: Rgb,
    success: Rgb,
    warning: Rgb,
    err: Rgb,
    info: Rgb,
    diff_add: Rgb,
    diff_delete: Rgb,

    pub fn dim(self: *const Theme) Rgb {
        return blend(self.foreground, self.background, 40);
    }

    pub fn codeFg(self: *const Theme) Rgb {
        return self.accent;
    }

    pub fn userBubbleBg(self: *const Theme) Rgb {
        return darken(self.success, 60);
    }

    pub fn assistantBubbleBg(self: *const Theme) Rgb {
        return darken(self.primary, 60);
    }

    pub fn codeBubbleBg(self: *const Theme) Rgb {
        return darken(self.assistantBubbleBg(), 20);
    }

    pub fn bubbleFg(self: *const Theme) Rgb {
        _ = self;
        return .{ 240, 240, 240 };
    }

    pub fn statusBarBg(self: *const Theme) Rgb {
        return self.primary;
    }

    pub fn statusBarFg(_: *const Theme) Rgb {
        return .{ 0, 0, 0 };
    }

    pub fn cursorBg(self: *const Theme) Rgb {
        return self.foreground;
    }

    pub fn cursorFg(self: *const Theme) Rgb {
        return self.background;
    }

    pub fn userLabelColor(self: *const Theme) Rgb {
        return self.success;
    }

    pub fn assistantLabelColor(self: *const Theme) Rgb {
        return self.primary;
    }

    pub fn toolResultColor(self: *const Theme) Rgb {
        return self.err;
    }

    pub fn toolDefaultColor(self: *const Theme) Rgb {
        return blend(self.info, self.foreground, 60);
    }

    pub fn inputBorderColor(self: *const Theme) Rgb {
        return self.dim();
    }

    pub fn pickerCursorBg(self: *const Theme) Rgb {
        return self.foreground;
    }

    pub fn pickerCursorFg(self: *const Theme) Rgb {
        return self.background;
    }
};

pub const ThemeEntry = struct {
    name: []const u8,
    id: []const u8,
};

fn blend(a: Rgb, b: Rgb, pct: u8) Rgb {
    return .{
        @intCast((@as(u16, a[0]) * pct + @as(u16, b[0]) * (100 - pct)) / 100),
        @intCast((@as(u16, a[1]) * pct + @as(u16, b[1]) * (100 - pct)) / 100),
        @intCast((@as(u16, a[2]) * pct + @as(u16, b[2]) * (100 - pct)) / 100),
    };
}

fn darken(c: Rgb, pct: u8) Rgb {
    return .{
        @intCast(@as(u16, c[0]) * pct / 100),
        @intCast(@as(u16, c[1]) * pct / 100),
        @intCast(@as(u16, c[2]) * pct / 100),
    };
}

pub fn hexToRgb(hex: []const u8) ?Rgb {
    if (hex.len < 6) return null;
    const start: usize = if (hex[0] == '#') 1 else 0;
    const s = hex[start..];
    if (s.len < 6) return null;
    const r = std.fmt.parseInt(u8, s[0..2], 16) catch return null;
    const g = std.fmt.parseInt(u8, s[2..4], 16) catch return null;
    const b = std.fmt.parseInt(u8, s[4..6], 16) catch return null;
    return .{ r, g, b };
}

const embedded_themes = .{
    .{ "amoled", "AMOLED", @embedFile("themes/amoled.json") },
    .{ "aura", "Aura", @embedFile("themes/aura.json") },
    .{ "ayu", "Ayu", @embedFile("themes/ayu.json") },
    .{ "carbonfox", "Carbonfox", @embedFile("themes/carbonfox.json") },
    .{ "catppuccin", "Catppuccin", @embedFile("themes/catppuccin.json") },
    .{ "catppuccin-frappe", "Catppuccin Frappé", @embedFile("themes/catppuccin-frappe.json") },
    .{ "catppuccin-macchiato", "Catppuccin Macchiato", @embedFile("themes/catppuccin-macchiato.json") },
    .{ "cobalt2", "Cobalt2", @embedFile("themes/cobalt2.json") },
    .{ "cursor", "Cursor", @embedFile("themes/cursor.json") },
    .{ "dracula", "Dracula", @embedFile("themes/dracula.json") },
    .{ "everforest", "Everforest", @embedFile("themes/everforest.json") },
    .{ "flexoki", "Flexoki", @embedFile("themes/flexoki.json") },
    .{ "github", "GitHub", @embedFile("themes/github.json") },
    .{ "gruvbox", "Gruvbox", @embedFile("themes/gruvbox.json") },
    .{ "kanagawa", "Kanagawa", @embedFile("themes/kanagawa.json") },
    .{ "lucent-orng", "Lucent ORNG", @embedFile("themes/lucent-orng.json") },
    .{ "material", "Material", @embedFile("themes/material.json") },
    .{ "matrix", "Matrix", @embedFile("themes/matrix.json") },
    .{ "mercury", "Mercury", @embedFile("themes/mercury.json") },
    .{ "monokai", "Monokai", @embedFile("themes/monokai.json") },
    .{ "nightowl", "Night Owl", @embedFile("themes/nightowl.json") },
    .{ "nord", "Nord", @embedFile("themes/nord.json") },
    .{ "one-dark", "One Dark", @embedFile("themes/one-dark.json") },
    .{ "onedarkpro", "One Dark Pro", @embedFile("themes/onedarkpro.json") },
    .{ "orng", "ORNG", @embedFile("themes/orng.json") },
    .{ "osaka-jade", "Osaka Jade", @embedFile("themes/osaka-jade.json") },
    .{ "palenight", "Palenight", @embedFile("themes/palenight.json") },
    .{ "rosepine", "Rosé Pine", @embedFile("themes/rosepine.json") },
    .{ "shadesofpurple", "Shades of Purple", @embedFile("themes/shadesofpurple.json") },
    .{ "solarized", "Solarized", @embedFile("themes/solarized.json") },
    .{ "synthwave84", "Synthwave '84", @embedFile("themes/synthwave84.json") },
    .{ "tokyonight", "Tokyo Night", @embedFile("themes/tokyonight.json") },
    .{ "vercel", "Vercel", @embedFile("themes/vercel.json") },
    .{ "vesper", "Vesper", @embedFile("themes/vesper.json") },
    .{ "zenburn", "Zenburn", @embedFile("themes/zenburn.json") },
};

pub const theme_count = embedded_themes.len;

pub fn listAll() [theme_count]ThemeEntry {
    var entries: [theme_count]ThemeEntry = undefined;
    inline for (embedded_themes, 0..) |t, i| {
        entries[i] = .{ .id = t[0], .name = t[1] };
    }
    return entries;
}

pub fn getByName(name: []const u8) ?Theme {
    inline for (embedded_themes) |t| {
        if (std.mem.eql(u8, t[0], name)) {
            return parseThemeJson(t[0], t[1], t[2]);
        }
    }
    return null;
}

fn parseThemeJson(id: []const u8, display_name: []const u8, json_data: []const u8) ?Theme {
    const parsed = std.json.parseFromSlice(std.json.Value, std.heap.page_allocator, json_data, .{}) catch return null;
    defer parsed.deinit();

    const root = parsed.value.object;
    const dark = (root.get("dark") orelse return null).object;
    const palette = (dark.get("palette") orelse return null).object;

    _ = display_name;

    return .{
        .name = id,
        .background = hexToRgb(strVal(palette.get("neutral"))) orelse return null,
        .foreground = hexToRgb(strVal(palette.get("ink"))) orelse return null,
        .primary = hexToRgb(strVal(palette.get("primary"))) orelse return null,
        .accent = hexToRgb(strVal(palette.get("accent"))) orelse return null,
        .success = hexToRgb(strVal(palette.get("success"))) orelse return null,
        .warning = hexToRgb(strVal(palette.get("warning"))) orelse return null,
        .err = hexToRgb(strVal(palette.get("error"))) orelse return null,
        .info = hexToRgb(strVal(palette.get("info"))) orelse return null,
        .diff_add = hexToRgb(strVal(palette.get("diffAdd"))) orelse return null,
        .diff_delete = hexToRgb(strVal(palette.get("diffDelete"))) orelse return null,
    };
}

fn strVal(v: ?std.json.Value) []const u8 {
    return switch (v orelse return "") {
        .string => |s| s,
        else => "",
    };
}

pub fn isPath(name: []const u8) bool {
    if (name.len == 0) return false;
    if (name[0] == '/' or name[0] == '~') return true;
    if (std.mem.endsWith(u8, name, ".json")) return true;
    return false;
}

pub fn loadFromFile(io: std.Io, allocator: std.mem.Allocator, path: []const u8) ?Theme {
    const data = std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .limited(64 * 1024)) catch return null;
    defer allocator.free(data);
    return parseThemeJson("custom", "Custom", data);
}

pub fn default() Theme {
    return .{
        .name = "default",
        .background = .{ 0, 0, 0 },
        .foreground = .{ 220, 220, 220 },
        .primary = .{ 120, 160, 255 },
        .accent = .{ 220, 190, 130 },
        .success = .{ 130, 220, 130 },
        .warning = .{ 255, 200, 80 },
        .err = .{ 220, 100, 100 },
        .info = .{ 150, 170, 200 },
        .diff_add = .{ 100, 200, 100 },
        .diff_delete = .{ 220, 100, 100 },
    };
}

// ---- Tests ----

test "hexToRgb parses valid hex" {
    const c = hexToRgb("#ff8040").?;
    try std.testing.expectEqual(@as(u8, 255), c[0]);
    try std.testing.expectEqual(@as(u8, 128), c[1]);
    try std.testing.expectEqual(@as(u8, 64), c[2]);
}

test "hexToRgb rejects short string" {
    try std.testing.expect(hexToRgb("#fff") == null);
}

test "getByName returns dracula" {
    const t = getByName("dracula") orelse return error.TestUnexpectedResult;
    try std.testing.expectEqualStrings("dracula", t.name);
    try std.testing.expect(t.background[0] < 50);
}

test "getByName returns null for unknown" {
    try std.testing.expect(getByName("nonexistent") == null);
}

test "listAll returns all themes" {
    const all = listAll();
    try std.testing.expectEqual(theme_count, all.len);
    try std.testing.expectEqualStrings("amoled", all[0].id);
}

test "default theme matches original hardcoded colors" {
    const d = default();
    try std.testing.expectEqual(@as(u8, 130), d.success[0]);
    try std.testing.expectEqual(@as(u8, 220), d.success[1]);
}

test "isPath detects file paths" {
    try std.testing.expect(isPath("/home/user/theme.json"));
    try std.testing.expect(isPath("~/mytheme.json"));
    try std.testing.expect(isPath("custom.json"));
    try std.testing.expect(!isPath("dracula"));
    try std.testing.expect(!isPath("tokyonight"));
    try std.testing.expect(!isPath(""));
}

File: src/tui/theme.rs
use ratatui::style::Color;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone)]
pub struct Theme {
    pub background: Color,
    pub foreground: Color,
    pub primary: Color,
    pub accent: Color,
    pub success: Color,
    pub warning: Color,
    pub error: Color,
    pub info: Color,
    pub diff_add: Color,
    pub diff_delete: Color,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThemeEntry {
    pub id: String,
    pub name: String,
}

#[derive(Debug, Deserialize)]
struct ThemeJson {
    id: String,
    name: String,
    background: String,
    foreground: String,
    primary: String,
    accent: String,
    success: String,
    warning: String,
    error: String,
    info: String,
    diff_add: String,
    diff_delete: String,
}

impl Theme {
    pub fn dim(&self) -> Color {
        Self::blend(self.foreground, self.background, 0.5)
    }

    pub fn code_fg(&self) -> Color {
        self.accent
    }

    pub fn user_bubble_bg(&self) -> Color {
        Self::darken(self.primary, 0.7)
    }

    pub fn selection_bg(&self) -> Color {
        Self::blend(self.accent, self.background, 0.35)
    }

    pub fn selection_fg(&self) -> Color {
        self.foreground
    }

    pub fn separator(&self) -> Color {
        Self::blend(self.foreground, self.background, 0.75)
    }

    pub fn header_bg(&self) -> Color {
        Self::blend(self.foreground, self.background, 0.88)
    }

    pub fn input_bg(&self) -> Color {
        Self::blend(self.primary, self.background, 0.90)
    }

    pub fn tool_border(&self) -> Color {
        Self::blend(self.info, self.background, 0.55)
    }

    pub fn blend(a: Color, b: Color, t: f64) -> Color {
        let (ar, ag, ab) = color_rgb(a);
        let (br, bg, bb) = color_rgb(b);
        Color::Rgb(lerp(ar, br, t), lerp(ag, bg, t), lerp(ab, bb, t))
    }

    pub fn darken(c: Color, factor: f64) -> Color {
        let (r, g, b) = color_rgb(c);
        Color::Rgb(
            (r as f64 * factor) as u8,
            (g as f64 * factor) as u8,
            (b as f64 * factor) as u8,
        )
    }
}

fn lerp(a: u8, b: u8, t: f64) -> u8 {
    (a as f64 + (b as f64 - a as f64) * t) as u8
}

fn color_rgb(c: Color) -> (u8, u8, u8) {
    match c {
        Color::Rgb(r, g, b) => (r, g, b),
        _ => (128, 128, 128),
    }
}

fn hex_to_rgb(hex: &str) -> Option<Color> {
    let hex = hex.strip_prefix('#').unwrap_or(hex);
    if hex.len() != 6 {
        return None;
    }
    let r = u8::from_str_radix(&hex[0..2], 16).ok()?;
    let g = u8::from_str_radix(&hex[2..4], 16).ok()?;
    let b = u8::from_str_radix(&hex[4..6], 16).ok()?;
    Some(Color::Rgb(r, g, b))
}

fn theme_from_json(j: &ThemeJson) -> Option<Theme> {
    Some(Theme {
        background: hex_to_rgb(&j.background)?,
        foreground: hex_to_rgb(&j.foreground)?,
        primary: hex_to_rgb(&j.primary)?,
        accent: hex_to_rgb(&j.accent)?,
        success: hex_to_rgb(&j.success)?,
        warning: hex_to_rgb(&j.warning)?,
        error: hex_to_rgb(&j.error)?,
        info: hex_to_rgb(&j.info)?,
        diff_add: hex_to_rgb(&j.diff_add)?,
        diff_delete: hex_to_rgb(&j.diff_delete)?,
    })
}

const BUILTIN_THEMES_JSON: &str = include_str!("themes.json");

fn load_builtin_themes() -> Vec<(ThemeEntry, Theme)> {
    let themes: Vec<ThemeJson> = serde_json::from_str(BUILTIN_THEMES_JSON).unwrap_or_default();
    themes
        .iter()
        .filter_map(|j| {
            let entry = ThemeEntry {
                id: j.id.clone(),
                name: j.name.clone(),
            };
            let theme = theme_from_json(j)?;
            Some((entry, theme))
        })
        .collect()
}

pub fn list_all() -> Vec<ThemeEntry> {
    let mut entries: Vec<ThemeEntry> = load_builtin_themes().into_iter().map(|(e, _)| e).collect();
    entries.sort_by_key(|a| a.name.to_lowercase());
    entries
}

pub fn get_by_name(name: &str) -> Option<Theme> {
    load_builtin_themes()
        .into_iter()
        .find(|(e, _)| e.id == name || e.name.eq_ignore_ascii_case(name))
        .map(|(_, t)| t)
}

pub fn default_theme() -> Theme {
    get_by_name("dracula").unwrap_or(Theme {
        background: Color::Rgb(40, 42, 54),
        foreground: Color::Rgb(248, 248, 242),
        primary: Color::Rgb(139, 233, 253),
        accent: Color::Rgb(189, 147, 249),
        success: Color::Rgb(80, 250, 123),
        warning: Color::Rgb(241, 250, 140),
        error: Color::Rgb(255, 85, 85),
        info: Color::Rgb(139, 233, 253),
        diff_add: Color::Rgb(80, 250, 123),
        diff_delete: Color::Rgb(255, 85, 85),
    })
}

pub fn load_from_file(path: &std::path::Path) -> Option<Theme> {
    let content = std::fs::read_to_string(path).ok()?;
    let j: ThemeJson = serde_json::from_str(&content).ok()?;
    theme_from_json(&j)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hex_to_rgb_valid() {
        assert_eq!(hex_to_rgb("#ff0000"), Some(Color::Rgb(255, 0, 0)));
        assert_eq!(hex_to_rgb("00ff00"), Some(Color::Rgb(0, 255, 0)));
    }

    #[test]
    fn hex_to_rgb_invalid() {
        assert_eq!(hex_to_rgb("xyz"), None);
        assert_eq!(hex_to_rgb("#gggggg"), None);
    }

    #[test]
    fn blend_midpoint() {
        let result = Theme::blend(Color::Rgb(0, 0, 0), Color::Rgb(100, 100, 100), 0.5);
        assert_eq!(result, Color::Rgb(50, 50, 50));
    }

    #[test]
    fn darken_factor() {
        let result = Theme::darken(Color::Rgb(200, 100, 50), 0.5);
        assert_eq!(result, Color::Rgb(100, 50, 25));
    }

    #[test]
    fn default_theme_exists() {
        let t = default_theme();
        assert!(matches!(t.background, Color::Rgb(_, _, _)));
    }

    #[test]
    fn builtin_themes_load() {
        let themes = list_all();
        assert!(!themes.is_empty());
    }

    #[test]
    fn get_by_name_found() {
        let themes = list_all();
        if let Some(first) = themes.first() {
            assert!(get_by_name(&first.id).is_some());
        }
    }
}
